### PR TITLE
Order update attributes refactor

### PR DESCRIPTION
**Pull request by jhawthorn**

### CODE REVIEW

**Comment by jordan-brough** [comment]
Looking great to me!  The extraction of code into standalone classes seems like what we've been needing and the spec cleanups along the way are great to have also.

File: core/app/models/spree/order_update_attributes.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    def initialize(order, attributes, request_env: nil, request: nil)
+      @order = order
+      @attributes = attributes.dup
+      @payments_attributes = @attributes.delete(:payments_attributes) || []
+      @request = request
+      @request_env = request_env
+      @request_env ||= request.headers.env if request
+    end
+
+    def update
```
Comment (jordan-brough) [inline core/app/models/spree/order_update_attributes.rb]: What do you think of `apply` instead of `update` for the name of this method?  Or something else like that.  If I have a "FooAttributes" object, I kind expect `foo_attributes.update` to refresh the attributes themselves or something like that.

Comment (jhawthorn) [inline core/app/models/spree/order_update_attributes.rb]: I like `apply`

**Comment by jhawthorn** [comment]
Disappointing discovery: if I comment out set_shipments_cost, all specs still pass :disappointed:. Will have to try and add a feature test exposing that.

File: api/app/controllers/spree/api/checkouts_controller.rb [hunk 2]
```diff
@@ -84,6 +77,33 @@ def user_id
           params[:order][:user_id] if params[:order]
         end
 
+        def update_params
+          if update_params = massaged_params[:order]
```
Comment (grzlus) [inline api/app/controllers/spree/api/checkouts_controller.rb]: I don't like setting variables in conditional statements. I propose 2 different wat how to handle it:

~~~
return {} if masseged_params[:order].blank?
massaged_params[:order].permit(permitted_checkout_attributes)
~~~

~~~
update_params = massaged_params.fetch(:order, {})
update_params.permit(permitted_checkout_attributes)
~~~

File: core/app/models/spree/order_update_attributes.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        false
+      end
+    end
+
+    private
+
+    attr_reader :attributes, :payments_attributes, :order
+
+    def assign_order_attributes
+      order.attributes = attributes
```
Comment (grzlus) [inline core/app/models/spree/order_update_attributes.rb]: I think, that explicit call `assign_attributes` will look better there.

File: core/app/models/spree/payment_create.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        payment.source = payment_method.payment_source_class.new(source_attributes)
+        payment.source.payment_method_id = payment_method.id
+        payment.source.user_id = order.user_id if order
+      end
+    end
+
+    def build_existing_card
+      credit_card = CreditCard.
+        where(user_id: order.user_id).
+        where.not(user_id: nil).
```
Comment (grzlus) [inline core/app/models/spree/payment_create.rb]: You don't need to check if `user_id is not null`. If user is null for order, then CreditCard will be not found, so `find` will throw exception.

In case when user_id is null, you can just stop executing code and throw exception. I don't know if there could be order without user at all.

Comment (jhawthorn) [inline core/app/models/spree/payment_create.rb]: Good catch. I think I'll extract this to another method and use `Spree::CreditCard.none`

Comment (grzlus) [inline core/app/models/spree/payment_create.rb]: Using `NullRelation` is a good idea, but should we proceed that method, when there will be no card at all?

Comment (jordan-brough) [inline core/app/models/spree/payment_create.rb]: @grzlus yes you can have orders with null `user_id` values.  I wanted to raise the same `ActiveRecord::RecordNotFound` error in either case so I put the condition in the query.  You could equivalently do something like `raise ActiveRecord::RecordNotFound if order.user_id.nil?` but I don't see any real benefit from that. Am I missing something?

File: core/app/models/spree/order_update_attributes.rb (hunk 3 above)
Comment (jordan-brough) [inline core/app/models/spree/order_update_attributes.rb]: In general I also prefer `assign_attributes` over `attributes=` since I find the latter to be a little misleading, but I'm not sure it'd be worth the time it would take for john to retype it. :) (Personally I like to defer to the author of the code on minor stylistic stuff.)

File: api/app/controllers/spree/api/checkouts_controller.rb (hunk 2 above)
Comment (jordan-brough) [inline api/app/controllers/spree/api/checkouts_controller.rb]: Seems a bit bike-shedding-ish...

Also, in this case I like how the current form calls out the fact that we may want to discontinue supporting requests without parameters (which I find weird).

File: core/app/models/spree/payment_create.rb (hunk 4 above)
Comment (grzlus) [inline core/app/models/spree/payment_create.rb]: @jordan-brough When you raise this exception by yourself then you don't need to perform query.

Of course it's better to define custom Exception, in other part you can catch your exception, not general `ActiveRecord::RecordNotFound` - because, what record is missing (user, card, shop)? Custom exceptions are good.

Comment (jordan-brough) [inline core/app/models/spree/payment_create.rb]: > you don't need to perform query

The query ends up being `where user_id is null and user_id is not null` and on top of that this is an indexed field. I can't imagine performing the query will be a performance problem.

> Custom exceptions are good

My thought was that we don't necessarily want to expose extra information here, security-wise. The old code does raise a custom exception, but it wasn't very informative and I'm not sure it was helpful.  To me it seems reasonable to say that this credit card id doesn't exist for the given user, period.  But I'm not 100% sold on that.

Comment (grzlus) [inline core/app/models/spree/payment_create.rb]: The query ends up being where user_id is null and user_id is not null and on top of that this is an indexed field. I can't imagine performing the query will be a performance problem.

But this is still a query, that you can avoid. In production env you can have DB on different machine, so there is not served via Unix Socket.

File: core/lib/spree/core/controller_helpers/payment_parameters.rb [hunk 5]
```diff
@@ -63,5 +61,106 @@ def move_payment_source_into_payments_attributes(original_params)
...
+    #        {
+    #          ...params...
+    #          amount: <the order total>,
+    #        },
+    #      ],
+    #      ...other params...
+    #    },
+    #    ...other params...
+    #  }
+    #
```
Comment (hhff) [inline core/lib/spree/core/controller_helpers/payment_parameters.rb]: these comments are my favorite part of this PR

File: core/app/models/spree/payment_create.rb (hunk 4 above)
Comment (jordan-brough) [inline core/app/models/spree/payment_create.rb]: Yes, what you say is completely true. I can imagine it making a 0.0001% difference in response time for real-world requests hitting this code.  Personally I like having all the logic in the same place, in the query, but I'm OK if someone wants to change that in this particular case.  What I'm against is a culture where we waste time on changing stuff like this with micro-optimizations when we have much bigger fish to fry.

Comment (jhawthorn) [inline core/app/models/spree/payment_create.rb]: I still intend to switch this to `CreditCard.none` (mostly because I think that's a little more clear). That will perform no query, so I'm not sure what this discussion is about.

Comment (grzlus) [inline core/app/models/spree/payment_create.rb]: @… As I siad, `NullRelation` is a good idea. The whole conversation is about "Does 1 query matters".

@jordan-brough When your database is is overloaded, one simple query like:

~~~ SQL
SELECT 1 FROM users WHERE email = "…"
~~~

This query (email has index) could take 5-30s. It happens, so query with contraction in clauses probably be also delayed, because there is no resources to process it.

Comment (jordan-brough) [inline core/app/models/spree/payment_create.rb]: @grzlus yes that is absolutely true, but if your database is overloaded to that extent and you are running the code in this area then you are already doomed and a micro-optimization like this isn't going to help.

I'm still fine w/ refactoring this, and `CreditCard.none` seems like a great option there.

File: api/app/controllers/spree/api/checkouts_controller.rb (hunk 2 above)
Comment (magnusvk) [inline api/app/controllers/spree/api/checkouts_controller.rb]: Yeah I'm with @jordan-brough and @jhawthorn on this. I think the assignment in the conditional makes a lot of sense here.

File: core/app/models/spree/order_update_attributes.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+module Spree
+  class OrderUpdateAttributes
```
Comment (magnusvk) [inline core/app/models/spree/order_update_attributes.rb]: Extracting this into its own class like this seems like a lot of boilerplate for the benefit of extracting just about 10 lines of code, no? Or are we intending to add more stuff to this class later?

Comment (jordan-brough) [inline core/app/models/spree/order_update_attributes.rb]: My understanding was that we'll add more stuff here later but we're starting out w/ the minimum that we need to do the payments stuff that triggered all this.

**Comment by jordan-brough** [comment]
This is looking pretty great! @jhawthorn do you have much more on your todo list here or do you think this is pretty close to ready?

**Comment by jhawthorn** [comment]
@jordan-brough I have just two things I want to check up on. The first you might be able to help clarify

Here https://github.com/solidusio/solidus/pull/357/files#diff-0a2f4b241f8427ca90d58eec529e0e8aR53 we assign `credit_card.verification_value` and there are tests which verify (though expectations) that that is assigned. However that record is never saved, and I see no place that that would be used. If we need that value to be around at some point we need to test for it.

Second issue is to find out why we have `set_shipments_cost`

**Comment by jhawthorn** [comment]
I've added a test for `set_shipments_cost`. Other than my question to you @jordan-brough I believe this is ready

**Comment by jordan-brough** [comment]
@jhawthorn taking a look now

**Comment by jordan-brough** [comment]
@jhawthorn I can't see a defininte use for the verification_value with existing cards.  We definitely haven't use it ourselves.

You probably already saw this, but [this is the commit](https://github.com/solidusio/solidus/commit/bbdead3b8ad9c4f364f702ed26afb0b2de3a51c5) that added this via the `cvc_confirm` parameter (which we now translate into `verification_value` in `controller_helpers/payment_parameters.rb` to be more consistent in what we call it).  I couldn't find any corresponding PR or anything but it does seem to be meant specifically for use with existing cards.

I'd guess that someone was requiring their customers to re-supply the CVV before being able to re-use an existing card, and that the in-memory credit card was making its way to the payment processor code (e.g. in code like [this](https://github.com/solidusio/solidus/blob/80cd52f066e3e98947627f54f5672686790cbf75/core/app/models/spree/payment.rb#L232)) and that code expected the verification value to be there.

Summary: I don't know.  When I moved that code around I was just trying to be extra cautious. I would be willing to bet that there are very few stores that use that param, if any.

**Comment by jhawthorn** [comment]
@jordan-brough thanks. I was mistaken and assumed that was added by us. Might as well leave it in for now, changing that would be ouside this PR's scope.

As of 3a5e16ab49687c7960d90bfbf8e05b09eff0ecd2 (which was horrible to track down) this is good to go by me. :ship:

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 7]
```diff
@@ -52,6 +51,26 @@ def update
     end
 
     private
+
+      def update_params
```
Comment (magnusvk) [inline frontend/app/controllers/spree/checkout_controller.rb]: I'm still not a fan of the fact that we define this twice even though we already have a module in place which we can use to DRY this up.

Comment (jordan-brough) [inline frontend/app/controllers/spree/checkout_controller.rb]: I'm still not a fan of changing this. :)

**Comment by jordan-brough** [comment]
@jhawthorn super! I'm going to take another look today and verify that it'll give us what we were looking for originally.

File: core/app/models/spree/order_update_attributes.rb (hunk 6 above)
Comment (magnusvk) [inline core/app/models/spree/order_update_attributes.rb]: @jhawthorn would be nice to hear your take here.

File: core/app/models/spree/order_update_attributes.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        false
+      end
+    end
+
+    private
+
+    attr_reader :attributes, :payments_attributes, :order
+
+    def assign_order_attributes
+      order.assign_attributes attributes
```
Comment (magnusvk) [inline core/app/models/spree/order_update_attributes.rb]: Can we just leave this in `apply`? I don't see the point of extracting one line of code that we only call once.

File: core/app/models/spree/payment_create.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
+module Spree
+  # Service object for creating new payments on an Order
+  class PaymentCreate
```
Comment (magnusvk) [inline core/app/models/spree/payment_create.rb]: This class I like a lot.

**Comment by magnusvk** [comment]
This looks great overall. Keen to still chat a bit more about `OrderUpdateAttributes`.

File: core/app/models/spree/payment_create.rb (hunk 9 above)
Comment (BenMorganIO) [inline core/app/models/spree/payment_create.rb]: lol same

File: api/app/controllers/spree/api/checkouts_controller.rb [hunk 10]
```diff
@@ -56,14 +56,7 @@ def complete
...
 
-        update_params = if params[:payment_source].present?
-          ActiveSupport::Deprecation.warn("Passing payment_source is deprecated. Send source parameters inside payments_attributes[:source_attributes].", caller)
-          move_payment_source_into_payments_attributes(params)
-        else
-          params
-        end
-
-        if @order.update_from_params(update_params, permitted_checkout_attributes, request.headers.env)
+        if OrderUpdateAttributes.new(@order, update_params, request_env: request.headers.env).apply
```
Comment (jordan-brough) [inline api/app/controllers/spree/api/checkouts_controller.rb]: @jhawthorn should this be a configurable class? e.g. `Spree::Config.order_update_attributes_class`

Comment (jordan-brough) [inline api/app/controllers/spree/api/checkouts_controller.rb]: and same question for `PaymentCreate`

File: core/app/models/spree/order_update_attributes.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
+module Spree
+  class OrderUpdateAttributes
+    # @param order [Order] existing (persisted) order
+    # @param attributes [Hash] attributes which are assigned to the new order.
+    #   These attributes should already have been filtered.
+    #   * :payments_attributes attributes
+    def initialize(order, attributes, request_env: nil)
```
Comment (jordan-brough) [inline core/app/models/spree/order_update_attributes.rb]: @jhawthorn can we add `current_user: nil` to this?  (And supply it when we call this). It would be really nice for logging and auditing to have that.

File: api/app/controllers/spree/api/checkouts_controller.rb (hunk 10 above)
Comment (jhawthorn) [inline api/app/controllers/spree/api/checkouts_controller.rb]: I think that might be a good idea but is outside the scope of this PR

Comment (jordan-brough) [inline api/app/controllers/spree/api/checkouts_controller.rb]: I think before we merge this PR it ought to be in a state that satisfies the original extension goal that we had in mind. Maybe we should chat sometime about how we could best do that with the current state of this PR?

File: core/app/models/spree/order_update_attributes.rb (hunk 11 above)
Comment (jhawthorn) [inline core/app/models/spree/order_update_attributes.rb]: Maybe that's the best way to do this, but I'm not sure. Can we discuss in another PR

Comment (jordan-brough) [inline core/app/models/spree/order_update_attributes.rb]: Per our chat we're good w/ this idea and I'll submit it as a follow-up once we merge this. Thanks!

File: api/app/controllers/spree/api/checkouts_controller.rb (hunk 10 above)
Comment (jordan-brough) [inline api/app/controllers/spree/api/checkouts_controller.rb]: I'll submit this also as a follow-up after this PR.

**Comment by jordan-brough** [comment]
:+1: :ship: !

File: core/spec/lib/spree/core/controller_helpers/payment_parameters_spec.rb [hunk 12]
```diff
@@ -77,4 +78,103 @@ class FakeController < ApplicationController
...
+    let(:params) do
+      ActionController::Parameters.new(
+        order: {
+          payments_attributes: [{}],
+          other_order_param: 1,
+        },
+        other_param: 2,
+      )
+    end
+    let(:order) { create(:order_with_line_items, line_items_price: 101.00, line_items_count: 1, shipment_cost: 0) }
```
Comment (mtomov) [inline core/spec/lib/spree/core/controller_helpers/payment_parameters_spec.rb]: Sorry to be picky on lots of good work. 

This factory is very expensive, and I do not see why persisted data would be needed at all in a controller helper class. Maybe use `build_stubbed` or Order.new if possible.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 13]
```diff
@@ -52,6 +51,26 @@ def update
...
+          # We current allow update requests without any parameters in them.
+          {}
+        end
+      end
+
+      def massaged_params
+        massaged_params = params.deep_dup
+
+        move_payment_source_into_payments_attributes(massaged_params)
+        move_existing_card_into_payments_attributes(massaged_params)
```
Comment (mtomov) [inline frontend/app/controllers/spree/checkout_controller.rb]: I think the deprecation warnings here do not mirror the ones in the API controller.

I guess that goes along Magnus's point that there is a duplication of code, which is error prone.

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 14]
```diff
@@ -170,10 +170,59 @@ def spree_post_address
...
+          Spree::Order.class_eval do
+            insert_checkout_step :new_step, after: :payment
+          end
+        end
+
+        after do
+          Spree::Order.checkout_flow(&@old_checkout_flow)
+        end
+
+        let(:order) { create(:order_with_line_items) }
```
Comment (mtomov) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: Well, the point of extracting logic into another class is that you are building an interface, on which you can start *relying*. 

This test basically says, I do not rely on the `OrderUpdateAttributes`, so I will double check that everything is as expected there.

Other tests also double check, all that creates trouble .. slow tests, single error -> multiple test failures, etc. This talk [J.B. Rainsberger - Integrated Tests Are A Scam](https://vimeo.com/80533536) describes the problem well.

So, maybe to do an `expect(OrderUpdateAttributes).to receive(:new).with(...).and_return(MockOrderUpdater)` and then do an expectation for `#apply` on the `MockOrderUpdater`.

I understand that this #update method is very complicated, and maybe in this case it is better to leave it as it is, but wanted to point out the pattern for not so "do it all" methods.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 13 above)
Comment (jordan-brough) [inline frontend/app/controllers/spree/checkout_controller.rb]: @mtomov that is because it is *not* deprecated here and is not intended to be, which actually goes along with my point that this is only coincidental (and temporary) duplication. :)

File: core/app/models/spree/order_update_attributes.rb (hunk 6 above)
Comment (mtomov) [inline core/app/models/spree/order_update_attributes.rb]: I am also a bit lost on the purpose of this class. Although the apparent intention is to function as a converter of html parameters to the model fields, 3 such methods are outside of it in a controller helper.

If that was not the intention, but instead to do business logic on order & payments, then would you not be effectively taking logic out of the Checkout state machine and placing it here. Is that the intention?

Do you not think it will be better to separate the business logic from the persistence? Otherwise, you will be continuing on the Spree path of "every class/method is on its own" - let's parse and save everything just to be sure.

The name `OrderUpdateAttributes` is far from revealing. What else would one be updating on an order, if it is not the attributes? It handles payments, calls out to shipments .. I would rather suggest to call it `CheckoutProgressor`. 

But then again, I think that the full Checkout code logic has to be drawn out in with a UML sort of diagam, so that the message flow will become clear, and from the message flow - the needed classes will naturally take their place. 

Sandi Metz in her book [Practical Object-Oriented Design in Ruby](http://www.poodr.com/) talks about such message flow diagrams and about higher importance of messages over the classes.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 13 above)
Comment (mtomov) [inline frontend/app/controllers/spree/checkout_controller.rb]: ah, all right. Good then : )

File: frontend/spec/controllers/spree/checkout_controller_spec.rb (hunk 14 above)
Comment (jhawthorn) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: This regression test was added before `OrderUpdateAttributes` was created. It provides us value in testing a corner case that the original code accounted for. For this reason it provides value.

Comment (mtomov) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: I see. Good then.
